[RFC] Laying down building stone for more flexible ONNX export capabilities

mfuntowicz opened the pull request:
This PR aims at reworking the way the ONNX export tool work by introducing a static, checked description format to provide ONNX exporters (pt almost done, TF will follow) all the required knobs.

More specifically this PR introduces the following concepts:
- `OnnxConfig` dataclass which enforces a model to be supported to describe all the properties to generate proper export
- `OnnxVariable` namedtuple which describe a variables w.r.t  the name of the variable, shape and potentially how many time it's "repeated" => Useful for `past_keys` 

Test case was done initially for BART model, without `use_cache=True` supports. 

For the sake of completeness, dropping support for `use_cache=True` is currently needed because we have a double nested tuple at the core of the `past_keys` output structure which would require multiple level of dynamic axis, not currently supported by ONNX.

This might be something we can work on in the future, potentially introducing a ONNX compatible output structure getting rid of the nested tuples layout and activable from a config property (_to be discussed further later on_).

**Update 1:** 

- I managed to enable exporting with nested structures such as `past_key_values` for GPT2.
- Need to work on enabling the same for using such values as inputs to the model

Supported models: 
- [x] ALBERT
- [x] BART (with & without past)
- [x] BERT 
- [x] DistilBERT
- [ ] Longformer => I've support for this, but the exporting fails because of missing ops ... need investigations.
- [x] GPT2 (with & without past)
- [x] Roberta
- [x] T5
- [x] XLM-Roberta

mfuntowicz committed:
Laying down building stone for more flexible ONNX export capabilities

mfuntowicz committed:
Ability to provide a map of config key to override before exporting.

mfuntowicz committed:
Makes it possible to export BART with/without past keys.

mfuntowicz committed:
Supports simple mathematical syntax for OnnxVariable.repeated

mfuntowicz committed:
Effectively apply value override from onnx config for model

mfuntowicz committed:
Supports export with additional features such as with-past for seq2seq

mfuntowicz committed:
Store the output path directly in the args for uniform usage across.

mfuntowicz committed:
Make BART_ONNX_CONFIG_* constants and fix imports.

mfuntowicz committed:
Support BERT model.

mfuntowicz committed:
Use tokenizer for more flexibility in defining the inputs of a model.

mfuntowicz committed:
Add TODO as remainder to provide the batch/sequence_length as CLI args

mfuntowicz committed:
Enable optimizations to be done on the model.

mfuntowicz committed:
Enable GPT2 + past

mfuntowicz committed:
Improve model validation with outputs containing nested structures

mfuntowicz committed:
Enable Roberta

mfuntowicz committed:
Enable Albert

mfuntowicz committed:
Albert requires opset >= 12

mfuntowicz committed:
BERT-like models requires opset >= 12

mfuntowicz committed:
Remove double printing.

mfuntowicz committed:
Enable XLM-Roberta

mfuntowicz committed:
Enable DistilBERT

mfuntowicz committed:
Disable optimization by default

mfuntowicz committed:
Fix missing setattr when applying optimizer_features

mfuntowicz committed:
Add value field to OnnxVariable to define constant input (not from tokenizers)

mfuntowicz committed:
Add T5 support.

mfuntowicz committed:
Simplify model type retrieval

mfuntowicz committed:
Example exporting token_classification pipeline for DistilBERT.

mfuntowicz commented:
Example of potential command line to export `bert-base-cased` => 

`python3 -m transformers.onnx --model=bert-base-cased --features=default --optimize --optimizations-level=all onnx/bert-base-cased/`

mfuntowicz committed:
Refactoring to package `transformers.onnx`

LysandreJik committed:
Solve circular dependency & __main__

LysandreJik committed:
Remove unnecessary imports in `__init__`

LysandreJik committed:
Licences

Narsil reviewed (commented):
Late to the party, but just a suggestion,

Why do OnnxConfig does not take the real Config object as an argument ? It would make all string-like inference `$config.hidden_size` unnecessary anymore, no ?

mfuntowicz committed:
Use @Narsil's suggestion to forward the model's configuration to the ONNXConfig to avoid interpolation.

mfuntowicz committed:
Fix Copied from

mfuntowicz committed:
LongFormer ONNX config.

LysandreJik committed:
Onnx export v2 fixes (#12388)

* Tiny fixes
Remove `convert_pytorch` from onnxruntime-less runtimes
Correct reference to model

* Style

mfuntowicz committed:
Removed optimizations

mfuntowicz committed:
Remvoe bad merge relicas.

mfuntowicz committed:
Remove unused constants.

mfuntowicz committed:
Remove some deleted constants from imports.

mfuntowicz committed:
Fix unittest to remove usage of PyTorch model for onnx.utils.

mfuntowicz committed:
Fix distilbert export

mfuntowicz committed:
Enable ONNX export test for supported model.

mfuntowicz committed:
Style.

mfuntowicz committed:
Fix lint.

mfuntowicz committed:
Enable all supported default models.

mfuntowicz committed:
GPT2 only has one output

mfuntowicz committed:
Fix bad property name when overriding config.

mfuntowicz committed:
Added unittests and docstrings.

mfuntowicz committed:
Disable with_past tests for now.

mfuntowicz committed:
Enable outputs validation for default export.

mfuntowicz committed:
Remove graph opt lvls.

mfuntowicz committed:
Last commit with on-going past commented.

mfuntowicz committed:
Style.

mfuntowicz committed:
Disabled `with_past` for now

mfuntowicz committed:
Remove unused imports.

mfuntowicz committed:
Remove framework argument

mfuntowicz committed:
Remove TFPreTrainedModel reference

LysandreJik committed:
Add documentation

mfuntowicz commented:
Idea: Rename the `convert_pytorch` to `export` so we have the exact same hierarchy than PyTorch: 
- PyTorch: `torch.onnx.export`
- Transformers: `transformers.onnx.export`

wdyt?

LysandreJik commented:
That's a great idea!

LysandreJik committed:
Add onnxruntime tests to CircleCI

LysandreJik committed:
Add test

mfuntowicz committed:
Rename `convert_pytorch` to `export`

mfuntowicz commented:
@Narsil we moved forward on your suggestion, can you have a look _(one more time 😄)_ 🙏🏻

mfuntowicz committed:
Use OrderedDict for dummy inputs

mfuntowicz committed:
Revert "WIP Wav2Vec2"

This reverts commit f665efb04c92525c3530e589029f0ae7afdf603e.

mfuntowicz committed:
WIP Wav2Vec2

LysandreJik committed:
Style

mfuntowicz committed:
Use OrderedDict for I/O

mfuntowicz committed:
Style.

mfuntowicz committed:
Specify OrderedDict documentation.

LysandreJik committed:
Style :)

leoozy commented:
Hello, when we can use the transformers.onnx?

LysandreJik commented:
You already can when installing from source:
```
pip install git+https://github.com/huggingface/transformers
```

We'll do a release this week (probably Thursday or Friday) and it will be in a pypi release then.

kagrze commented on src/transformers/models/t5/configuration_t5.py:
```diff
@@ -132,3 +137,66 @@ def num_attention_heads(self):
+            [
+                ("input_ids", {0: "batch", 1: "encoder_sequence"}),
+                ("attention_mask", {0: "batch", 1: "encoder_sequence"}),
+                ("decoder_input_ids", {0: "batch"}),
```
I see that the `decoder_input_ids` length is fixed. I guess, this makes sense when `use_past` is `True`, because we feed only one token (the one generated in the previous step). However, when `use_past` is `False` then we have to feed all the previously generated tokens, haven't we?

Avi-avidan commented:
hi, this thread is super important. 
Is there support for bart text2text_generation export to onnx (more specifically for summarization tasks) ?
